hw/xbox/nv2a/nv2a_pgraph.c

```diff
@@ -299,26 +299,81 @@ static uint64_t fnv_hash(const uint8_t *data, size_t len);
 static uint64_t fast_hash(const uint8_t *data, size_t len, unsigned int samples);
 
 /* PGRAPH - accelerated 2d/3d drawing engine */
+
+static uint32_t pgraph_rdi_read(PGRAPHState *pg,
+                                unsigned int select, unsigned int address)
+{
+    uint32_t r = 0;
+    switch(select) {
+    case RDI_INDEX_VTX_CONSTANTS0:
+    case RDI_INDEX_VTX_CONSTANTS1:
+        assert((address / 4) < NV2A_VERTEXSHADER_CONSTANTS);
+        r = pg->vsh_constants[address / 4][3 - address % 4];
+        break;
+    default:
+        fprintf(stderr, "nv2a: unknown rdi read select 0x%x address 0x%x\n", select, address);
+        assert(false);
+        break;
+    }
+    return r;
+}
+
+static void pgraph_rdi_write(PGRAPHState *pg,
+                             unsigned int select, unsigned int address,
+                             uint32_t val)
+{
+    switch(select) {
+    case RDI_INDEX_VTX_CONSTANTS0:
+    case RDI_INDEX_VTX_CONSTANTS1:
+        assert(false); /* Untested */
+        assert((address / 4) < NV2A_VERTEXSHADER_CONSTANTS);
+        pg->vsh_constants_dirty[address / 4] |=
+            (val != pg->vsh_constants[address / 4][3 - address % 4]);
+        pg->vsh_constants[address / 4][3 - address % 4] = val;
+        break;
+    default:
+        fprintf(stderr, "nv2a: unknown rdi write select 0x%x, address 0x%x\n", select, address);
+        assert(false);
+        break;
+    }
+}
+
 uint64_t pgraph_read(void *opaque, hwaddr addr, unsigned int size)
 {
     NV2AState *d = (NV2AState *)opaque;
+    PGRAPHState *pg = &d->pgraph;
 
-    qemu_mutex_lock(&d->pgraph.lock);
+    qemu_mutex_lock(&pg->lock);
 
     uint64_t r = 0;
     switch (addr) {
     case NV_PGRAPH_INTR:
-        r = d->pgraph.pending_interrupts;
+        r = pg->pending_interrupts;
         break;
     case NV_PGRAPH_INTR_EN:
-        r = d->pgraph.enabled_interrupts;
+        r = pg->enabled_interrupts;
+        break;
+    case NV_PGRAPH_RDI_DATA: {
+        unsigned int select = GET_MASK(pg->regs[NV_PGRAPH_RDI_INDEX],
+                                       NV_PGRAPH_RDI_INDEX_SELECT);
+        unsigned int address = GET_MASK(pg->regs[NV_PGRAPH_RDI_INDEX],
+                                        NV_PGRAPH_RDI_INDEX_ADDRESS);
+
+        r = pgraph_rdi_read(pg, select, address);
+
+        /* FIXME: Overflow into select? */
+        assert(address < GET_MASK(NV_PGRAPH_RDI_INDEX_ADDRESS,
+                                  NV_PGRAPH_RDI_INDEX_ADDRESS));
```

Comment:
Untested; I might run a unit test for this in the next days.

(Other instances too)

Reply:
Got no time for this at the moment. Code can remain as-is and improved later; it will trigger an assert before unknown behaviour.